### PR TITLE
feat(cache): Compression for repository cache

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Compress repo cache data with `snappy` algorithm
- Migrate from previous revisions

## Context

- Ref: #15275

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
Compression rate is about x4

**Review by rarkins — DISMISSED** [review]

File: package.json [hunk 1]
```diff
@@ -203,6 +203,7 @@
     "shlex": "2.1.0",
     "simple-git": "3.6.0",
     "slugify": "1.6.5",
+    "snappy": "7.1.1",
```
Comment (viceice) [inline package.json:206]: can't we use some of our existing deps or node buildins? looks like a pretty huge dep and yarn will always install all optional depencies. 🤔

**Comment by viceice** [comment]
what about buildin zlib?
https://nodejs.org/api/zlib.html

**Comment by viceice** [comment]
or pure js snappy

https://github.com/zhipeng-jia/snappyjs

**Comment by rarkins** [comment]
Yes let's use snappyjs

**Comment by zharinov** [comment]
In this case, I think it makes sense to stick to built-in `gzip`

**Comment by zharinov** [comment]
My primary reason was to prefer speed over compression rate, but pure JS implementation is twice slower anyway

File: lib/util/cache/repository/impl/local.ts [hunk 2]
```diff
@@ -1,21 +1,30 @@
...
+
 export class LocalRepoCache extends RepoCacheBase {
+  private oldHash: string | null = null;
+
   constructor(private platform: string, private repository: string) {
     super();
   }
 
-  private getCacheFileName(): string {
+  public getCacheFileName(): string {
```
Comment (zharinov) [inline lib/util/cache/repository/impl/local.ts:27]: This one is intentional

**Comment by zharinov** [comment]
For Brotli algorithm, compression rate is `x8`. Also I've added check that would short-circuit if the data isn't changed.

**Review by rarkins — DISMISSED** [review]

File: lib/util/cache/repository/impl/local.spec.ts [hunk 3]
```diff
@@ -6,6 +9,21 @@ import { LocalRepoCache } from './local';
...
 
+const compress = promisify(zlib.brotliCompress);
+
+async function createCacheRecord(
+  data: RepoCacheData,
+  repository = 'some/repo'
+): Promise<RepoCacheRecord> {
+  const revision = CACHE_REVISION;
+  const jsonStr = JSON.stringify(data);
+  const hash = hasha(jsonStr, { algorithm: 'sha1' });
```
Comment (viceice) [inline lib/util/cache/repository/impl/local.spec.ts:20]: shouldn't we use sha256 to be more safe?

File: lib/util/cache/repository/impl/local.ts [hunk 4]
```diff
@@ -50,7 +74,13 @@ export class LocalRepoCache extends RepoCacheBase {
     const revision = CACHE_REVISION;
     const repository = this.repository;
     const data = this.getData();
-    const record: RepoCacheRecord = { revision, repository, data };
-    await outputFile(cacheFileName, JSON.stringify(record));
+    const jsonStr = JSON.stringify(data);
+    const hash = await hasha.async(jsonStr, { algorithm: 'sha1' });
```
Comment (viceice) [inline lib/util/cache/repository/impl/local.ts:78]: as above

**Review by viceice — COMMENTED** [review]
otherwise LGTM.

can you check that the new code and also tests are null safe?

File: lib/util/cache/repository/impl/local.spec.ts (hunk 3 above)
Comment (zharinov) [reply lib/util/cache/repository/impl/local.spec.ts:20]: I believe it has nothing to do with encryption/security, I'm just using it the similar way it's being used in `git`

Comment (viceice) [reply lib/util/cache/repository/impl/local.spec.ts:20]: git is also slowly transitioning to sha256 😉

Comment (zharinov) [reply lib/util/cache/repository/impl/local.spec.ts:20]: Haha, you got me

**Review by rarkins — APPROVED** [review]
